File: BinarySearch/Python/BinarySearch.py
import random
import math

def binarySearch(arr, value):
   
    print(arr)
    if(len(arr) == 1 and value != arr[0]):
        print("%d wasnt found" % value)
        return -1
    else:

        middleIndex = math.floor(len(arr) / 2)
        middleValue = arr[middleIndex]
    
        if(middleValue == value):
            print("%d was found" % value)
            return 1

        else:
            if(value > middleValue):
                return binarySearch(arr[middleIndex:], value)
            elif(value < middleValue):
                return binarySearch(arr[:middleIndex], value)

arr = []
[arr.append(i*i) for i in range(30)]

#print(arr[:16])
#print(arr[16:])

print(arr)
print(binarySearch(arr, 2))
print(binarySearch(arr, 4))
print(binarySearch(arr, 10))
print(binarySearch(arr, 257))
print(binarySearch(arr, 256))
print(binarySearch(arr, 400))
print(binarySearch(arr, 441))

File: BinarySearch/Go/BinarySearch.go
package main

import (
	"fmt"
)

func binarySearch(slice []int, value int) int {

	var beginning = 0
	var end = len(slice) - 1
	
	for end >= beginning {

		var middleIndex = (beginning + end) / 2
		var middleValue = slice[middleIndex]

		if(value == middleValue) {
			
			fmt.Println(value, " was found")
			return middleIndex
		
		} else if (value < middleValue) {
			
			end = middleIndex - 1
		
		} else {
			
			beginning = middleIndex + 1
		
		}
	}

	fmt.Println(value, "was not found")
	return -1
}

func main() {

	slice := []int{}
	
	for i := 0; i < 30; i++ {
		slice = append(slice, i*i)
	}
	
	fmt.Println(slice)
	fmt.Println(binarySearch(slice, 2))
	fmt.Println(binarySearch(slice, 4))
	fmt.Println(binarySearch(slice, 10))
	fmt.Println(binarySearch(slice, 257))
	fmt.Println(binarySearch(slice, 256))
	fmt.Println(binarySearch(slice, 400))
	fmt.Println(binarySearch(slice, 441))
}
